### PR TITLE
Added File IO in Scala Article

### LINKED ISSUE #1031: Add Article for File IO in Scala

### PR DESCRIPTION
**Pull request by veevidify**
Congrats on taking the first step to contributing to the Sample Programs repository maintained by [The Renegade Coder][1]! For simplicity, please make sure that your pull request includes one and only one contribution.

## Complete the Applicable Sections Below

Find which section best describes your pull request and make sure you fill it out. To start, let us know which issue you've fixed.

- [x] I fixed #1031 

### Code Snippets

- [ ] I named the pull request using `Added/Updated <Sample Program> in <Language>` format
- [ ] I created/updated the language README
  - [ ] I added the sample program name to the README
  - [ ] I added fun facts (i.e. debut, developer, typing, etc.)
  - [ ] I added reference link(s) to the README
  - [ ] I added solution citations when necessary (see [plagiarism][2])

### Documentation

- [x] I named the pull request using `Added/Updated <Sample Program> in <Language> Article` format
- [x] I followed the applicable article template
  - [ ] I added/updated a [language article][4] (i.e. The Python Programming Language)
  - [x] I added/updated a [code snippet article][3] (i.e. Hello World in Perl)
  - [ ] I added/updated a [project article][5] (i.e. Fizz Buzz in Every Language)

### Testing

- [ ] I names the pull request using `Added/Updated <Language>/<Project> Testing` format
- [ ] I followed the testinfo template, if applicable

## Notes

Feel free to describe what you added or updated.

[1]: https://therenegadecoder.com/
[2]: https://github.com/TheRenegadeCoder/sample-programs/blob/master/.github/CONTRIBUTING.md#plagiarism
[3]: https://github.com/TheRenegadeCoder/sample-programs/blob/master/docs/templates/CODE_ARTICLE_TEMPLATE.md
[4]: https://github.com/TheRenegadeCoder/sample-programs/blob/master/docs/templates/LANGUAGE_ARTICLE_TEMPLATE.md
[5]: https://github.com/TheRenegadeCoder/sample-programs/blob/master/docs/templates/PROJECT_ARTICLE_TEMPLATE.md

### CODE REVIEW

**Comment by jrg94** [comment]
Great stuff! Hope to get around to reading this sometime today. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

File: docs/projects/file-io/_posts/2019-04-05-scala.md [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+title: File IO in Scala
+layout: default
+last-modified: 2019-04-05
+featured-image:
+tags: [scala, file-io]
+authors:
+  - veevidify
+---
+
+Most of the time, the solution to file IO in a programming language is pretty straightforward. However in this article, we'll go just a bit further, utilising Scala's seamless interopability with Java as well as the powerful built-in `Option` class.
```
Comment (jrg94) [inline docs/projects/file-io/_posts/2019-04-05-scala.md]: Can you limit the line length to like 120 characters or so? For example:

```suggestion
Most of the time, the solution to file IO in a programming language is pretty straightforward. 
However in this article, we'll go just a bit further, utilising Scala's seamless interopability with 
Java as well as the powerful built-in `Option` class.
```

File: docs/projects/file-io/_posts/2019-04-05-scala.md [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+  }
+}
+```
+
+### Read from a file
+
+In many real world scenario, your program usually takes one of the two approaches when dealing with undeterministic I/O (also called side-effects): fail loudly or fail silently.
+
+We want to fail loudly, i.e. throw exception, stop execution flow or output error as results, when the I/O is critical for the program to proceed.
+```scala
```
Comment (jrg94) [inline docs/projects/file-io/_posts/2019-04-05-scala.md]: Could you add some spaces between code and paragraphs:

```suggestion

```scala
```

File: docs/projects/file-io/_posts/2019-04-05-scala.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+val optionLines = readFromFileIntoOption("input.txt")
+optionLines match {
+  case Some(lines) => lines.foreach(println)
+  case None =>
+}
+```
+
+### Write to a file
+
+Similar pattern of with / without `Option` for writing to files, however both approach serve the same purpose. Unlike read, which is usually at the beginning of some logic and having the input type determined gains us benefits, write is usually a void function with no return types, and called at the end of execution (result, log, etc.).
```
Comment (jrg94) [inline docs/projects/file-io/_posts/2019-04-05-scala.md]:
```suggestion
Similar pattern of with / without `Option` for writing to files, however both approaches serve the same purpose. Unlike read, which is usually at the beginning of some logic and having the input type determined gains us benefits, write is usually a void function with no return types, and called at the end of execution (result, log, etc.).
```

**Comment by veevidify** [comment]
All these changes look fine to me. Thanks.

**Comment by jrg94** [comment]
If you want to include the changes, I think you have to pull them in. I've never been able to commit my own suggestions... maybe I'm not doing it right.

**Comment by veevidify** [comment]
Sorry didn't realise there are multiple instances of your suggestions. Pushed a new commit just now.

Also I believe you could go to the "Files changed" tab and batch all the changes into a commit (would be made by you).

**Comment by jrg94** [comment]
Looks great! Thanks again for your work.

**Comment by jrg94** [comment]
One thing I forgot about before I merged... Can you add yourself to the authors file? https://github.com/TheRenegadeCoder/sample-programs/blob/master/docs/_data/authors.yml

**Comment by veevidify** [comment]
You already merged, which is fine. I'll fine another chance to append to the authors list.

**Pull request merged**